worker/worker.go

```diff
@@ -570,6 +574,84 @@ func (w *worker) rhpBroadcastHandler(jc jape.Context) {
 	}
 }
 
+func (w *worker) rhpPruneContractHandlerPOST(jc jape.Context) {
+	// decode fcid
+	var id types.FileContractID
+	if jc.DecodeParam("id", &id) != nil {
+		return
+	}
+
+	// check if there's prunable data for the contract
+	ctx := jc.Request.Context()
+	n, err := w.bus.PrunableDataForContract(ctx, id)
+	if errors.Is(err, api.ErrContractNotFound) {
+		jc.Error(err, http.StatusNotFound)
+		return
+	} else if jc.Check("couldn't fetch prunable data for contract", err) != nil {
+		return
+	} else if n == 0 {
+		jc.Encode(n)
+		return
+	}
+
+	// attach gouging checker
+	gp, err := w.bus.GougingParams(ctx)
+	if jc.Check("could not get gouging parameters", err) != nil {
+		return
+	}
+	ctx = WithGougingChecker(ctx, w.bus, gp)
+
+	// fetch the contract from the bus
+	c, err := w.bus.Contract(ctx, id)
+	if errors.Is(err, api.ErrContractNotFound) {
+		jc.Error(err, http.StatusNotFound)
+		return
+	} else if jc.Check("couldn't fetch contract", err) != nil {
+		return
+	}
+
+	// convenience variables
+	hk := c.HostKey
+	hostIP := c.HostIP
+	hostAddr := c.SiamuxAddr
+
+	// prune sectors from contracts
+	var toDelete []uint64
+	if jc.Check("couldn't prune contract", w.withRevision(ctx, defaultRevisionFetchTimeout, id, hk, hostAddr, lockingPriorityPruning, 0, func(rev types.FileContractRevision) error {
+		// fetch the roots from the host
+		renterKey := w.deriveRenterKey(c.HostKey)
+		got, err := w.FetchContractRoots(ctx, hostIP, hk, renterKey, id, rev.RevisionNumber, time.Minute)
+		if err != nil {
+			return err
+		}
+
+		// fetch the roots from the bus
+		roots, err := w.bus.ContractRoots(ctx, id)
+		if err != nil {
+			return err
+		}
+		keep := make(map[types.Hash256]struct{})
+		for _, root := range roots {
+			keep[root] = struct{}{}
+		}
+
+		// collect indices for roots we want to prune
+		for i, root := range got {
+			if _, wanted := keep[root]; wanted {
+				delete(keep, root) // prevent duplicates
```

Comment:
Strictly speaking it is allowed for a root to exist twice within the same contract at different indices.

Reply:
might as well prune duplicates